src/gamma.jl

```diff
@@ -740,13 +740,21 @@ end
 
 Euler integral of the first kind ``\\operatorname{B}(x,y) = \\Gamma(x)\\Gamma(y)/\\Gamma(x+y)``.
 """
-function beta(a::Number, b::Number)
+beta(a::Number, b::Number) = _beta(map(float, promote(a, b))...)
+
+# here `T` is a floating point type but we don't want to restrict the implementation
+# to `AbstractFloat` or `Float64`
+function _beta(a::T, b::T) where {T<:Number}
+    # two special cases
+    a == 1 && return inv(b)
```

Comment:
I believe this will introduce wrong results ForwardDiff based AD along the lines discussed in https://github.com/JuliaDiff/ForwardDiff.jl/issues/480

Reply:
I guess this doesn't matter given https://github.com/JuliaDiff/DiffRules.jl/blob/master/src/rules.jl#L179-L180